helios-client/src/main/java/com/spotify/helios/client/Endpoints.java

```diff
@@ -0,0 +1,170 @@
+ */
+public class Endpoints {
+
+  private static final Logger log = LoggerFactory.getLogger(Endpoints.class);
+
+  private static final DnsResolver DEFAULT_DNS_RESOLVER = SystemDefaultDnsResolver.INSTANCE;
+
+  private Endpoints() {
+  }
+
+  /**
+   * Returns a {@link Supplier} of a list of {@link Endpoint}.
+   * @param uriSupplier A Supplier of a list of URIs.
+   * @return A Supplier of a list of Endpoints.
+   */
+  public static Supplier<List<Endpoint>> of(final Supplier<List<URI>> uriSupplier) {
+    return of(uriSupplier, DEFAULT_DNS_RESOLVER);
+  }
+
+  /**
+   * Returns a {@link Supplier} of a list of {@link Endpoint}.
+   * @param uriSupplier A Supplier of a list of URIs.
+   * @param dnsResolver An instance of {@link DnsResolver}
+   * @return A Supplier of a list of Endpoints.
+   */
+  static Supplier<List<Endpoint>> of(final Supplier<List<URI>> uriSupplier,
+                                     final DnsResolver dnsResolver) {
+    return new Supplier<List<Endpoint>>() {
+      @Override
+      public List<Endpoint> get() {
+        return of(uriSupplier.get(), dnsResolver);
+      }
+    };
+  }
+
+  /**
+   * Returns a list of {@link Endpoint}.
+   * @param uris A list of URIs.
+   * @return A list of Endpoints.
+   */
+  public static List<Endpoint> of(final List<URI> uris) {
+    return of(uris, DEFAULT_DNS_RESOLVER);
+  }
+
+  /**
+   * Returns a list of {@link Endpoint}.
+   * @param uris A list of URIs.
+   * @param dnsResolver An instance of {@link DnsResolver}
+   * @return A list of Endpoints.
+   */
+  static List<Endpoint> of(final List<URI> uris, final DnsResolver dnsResolver) {
+    final ImmutableList.Builder<Endpoint> endpoints = ImmutableList.builder();
+    for (final URI uri : uris) {
+      try {
+        for (final InetAddress ip : dnsResolver.resolve(uri.getHost())) {
+          endpoints.add(new DefaultEndpoint(uri, ip));
+        }
+      } catch (UnknownHostException e) {
+        log.warn("Unable to resolve hostname {} into IP address: {}", uri.getHost(), e);
+      }
+    }
+
+    return endpoints.build();
+  }
+
+  private static class DefaultEndpoint implements Endpoint {
+
+    private static final List<String> VALID_PROTOCOLS = ImmutableList.of("http", "https");
+    private static final String VALID_PROTOCOLS_STR =
+        String.format("[%s]", Joiner.on("|").join(VALID_PROTOCOLS));
+
+    private final InetAddress ip;
+    private final URI uri;
+
+    DefaultEndpoint(final URI uri, final InetAddress ip) {
+      this.uri = checkNotNull(uri);
+      this.ip = checkNotNull(ip);
+
+      final String scheme = this.uri.getScheme();
+      final String host = this.uri.getHost();
+      final int port = this.uri.getPort();
+      if (!VALID_PROTOCOLS.contains(scheme) || host == null || port == -1) {
+        throw new IllegalArgumentException(String.format(
+            "Master endpoints must be of the form \"%s://heliosmaster.domain.net:<port>\"",
```

Comment:
if we don't actually care what format the hostname has (which this error message kind of suggests we do), I'd make this message be `Endpoints must be of the form %s://<host>:<port>`